processor/src/main/java/com/linecorp/decaton/processor/metrics/Metrics.java

```diff
@@ -45,15 +52,55 @@ private Metrics(AvailableTags availableTags) {
         this.availableTags = availableTags;
     }
 
-    public class SubscriptionMetrics {
+    /**
+     * Superclass of all *Metrics classes managing lifecycle of meter instances.
+     * This class implements refcounting for meters identified by {@link Meter.Id} and make sure to remove
+     * the instance from {@link MeterRegistry} when the last reference to the meter closed and disappeared.
+     */
+    abstract static class AbstractMetrics implements AutoCloseable {
+        private static final Map<Meter.Id, AtomicInteger> meterRefCounts = new HashMap<>();
+        private final List<Meter> meters;
+
+        protected AbstractMetrics() {
+            meters = new ArrayList<>();
+        }
+
+        <T extends Meter> T meter(Supplier<T> ctor) {
+            final T meter;
+            synchronized (meterRefCounts) {
+                meter = ctor.get();
+                meterRefCounts.computeIfAbsent(meter.getId(), key -> new AtomicInteger())
+                              .incrementAndGet();
+            }
+            meters.add(meter);
+            return meter;
+        }
+
+        @Override
+        public void close() {
+            for (Meter meter : meters) {
+                meter.close();
```

Comment:
I'm not sure if it's safe to close the meter here, since the meter can be updated/queried after close until all references deleted.

As far as checking Meter impls in micrometer-core, it seems `close()` is not overridden in almost all cases so maybe it's safe to close here, but I think it's better to move `close()` to this line: https://github.com/line/decaton/pull/38/files#diff-78df474b860bfdbeae88ca6d9ebd1c17R88

Reply:
That's a good point. Actually I've also considered which scope should we expect as a lifecycle for a meter instance.
My thought is, since `XXBuilder.build()` (which internally calls `registry.counter, gauge, timer...`) transparently returns an existing object if already presents, the returned objective's `close` should be safe to be called w/o extra assumption against possibly shared other instances.
On the other hand, `registry.remove` is difference since we're operating directly on the registry and explicitly ordering it to "remove" the given ID from it which it makes sense for me to work as a global destruction.
Currently the close method seems to be really pointless (not doing anything) except few cases like Timer as you said, so it really doesn't matter, but wouldn't the above interpretation makes much sense?

Reply:
Hm that makes sense. Thank you for your explanation

Reply:
Hmm, after I looked into `AbstractTimer.close => PauseDetector.close` now I changed my mind that it would not be expected to call close multiple times from difference instances...
Let me fix in a way you suggested then.